src/transport/SecureSessionMgr.cpp

```diff
@@ -92,136 +92,154 @@ CHIP_ERROR SecureSessionMgr::SendMessage(NodeId peerNodeId, System::PacketBuffer
+
+exit:
+    return err;
 }
 
-CHIP_ERROR SecureSessionMgr::SendMessage(PayloadHeader & payloadHeader, NodeId peerNodeId, System::PacketBufferHandle msgBuf,
-                                         EncryptedPacketBufferHandle * bufferRetainSlot, bool isEncrypted)
+CHIP_ERROR SecureSessionMgr::EncryptMessage(Transport::PeerConnectionState * state, PayloadHeader & payloadHeader,
+                                            PacketHeader & packetHeader, NodeId peerNodeId, System::PacketBufferHandle & msgBuf)
 {
-    CHIP_ERROR err              = CHIP_NO_ERROR;
-    PeerConnectionState * state = nullptr;
-    PacketHeader packetHeader;
-    uint16_t headerSize = 0;
+    CHIP_ERROR err         = CHIP_NO_ERROR;
+    uint8_t * data         = nullptr;
+    uint32_t payloadLength = 0; // Make sure it's big enough to add two 16-bit ints without overflowing.
+    uint16_t totalLen      = 0;
+    uint16_t taglen        = 0;
+    uint16_t actualEncodedHeaderSize;
+    MessageAuthenticationCode mac;
+
+    uint32_t msgId = state->GetSendMessageIndex();
+
+    static_assert(std::is_same<decltype(msgBuf->TotalLength()), uint16_t>::value,
+                  "Addition to generate payloadLength might overflow");
+
+    uint16_t headerSize = payloadHeader.EncodeSizeBytes();
+    payloadLength       = static_cast<uint32_t>(headerSize + msgBuf->TotalLength());
+    VerifyOrExit(CanCastTo<uint16_t>(payloadLength), err = CHIP_ERROR_NO_MEMORY);
+
+    packetHeader
+        .SetSourceNodeId(mLocalNodeId)              //
+        .SetDestinationNodeId(peerNodeId)           //
+        .SetMessageId(msgId)                        //
+        .SetEncryptionKeyID(state->GetLocalKeyID()) //
+        .SetPayloadLength(static_cast<uint16_t>(payloadLength));
+    packetHeader.GetFlags().Set(Header::FlagValues::kSecure);
+
+    VerifyOrExit(msgBuf->EnsureReservedSize(headerSize), err = CHIP_ERROR_NO_MEMORY);
+
+    msgBuf->SetStart(msgBuf->Start() - headerSize);
+    data     = msgBuf->Start();
+    totalLen = msgBuf->TotalLength();
+
+    err = payloadHeader.Encode(data, totalLen, &actualEncodedHeaderSize);
+    SuccessOrExit(err);
 
-    if (isEncrypted)
-    {
-        err = packetHeader.Decode(msgBuf->Start(), msgBuf->DataLength(), &headerSize);
-        SuccessOrExit(err);
+    err = state->GetSecureSession().Encrypt(data, totalLen, data, packetHeader, mac);
+    SuccessOrExit(err);
+    err = mac.Encode(packetHeader, &data[totalLen], kMaxTagLen, &taglen);
+    SuccessOrExit(err);
+
+    VerifyOrExit(CanCastTo<uint16_t>(totalLen + taglen), err = CHIP_ERROR_INTERNAL);
+    msgBuf->SetDataLength(static_cast<uint16_t>(totalLen + taglen));
+
+    state->IncrementSendMessageIndex();
 
-        VerifyOrExit(packetHeader.GetDestinationNodeId().HasValue(), err = CHIP_ERROR_INVALID_DESTINATION_NODE_ID);
-        peerNodeId = packetHeader.GetDestinationNodeId().Value();
+    ChipLogDetail(Inet, "Secure transport encrypted msg %u", msgId);
+
+exit:
+    if (!msgBuf.IsNull())
```

Comment:
if (!msgBuf.IsNull()) is always true since msgBuf is not moved. Suggest to use err to check the error case
 
    if (err != CHIP_NO_ERROR)
    {
        const char * errStr = ErrorStr(err);
        ChipLogError(Inet, "Secure transport failed to encrypt msg %u: %s", state->GetSendMessageIndex(), errStr);
    }
    else
    {
        state->IncrementSendMessageIndex();
    }